working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+  ShareableStream<E> toShareable() => /* ... */;
+}
+```
+
+The reason for providing these is to allow developers to structure their code
+using well understood primitives futures and streams instead of devising new
+primitives which reimplement some of the same functionality and is compatible
+with threading.
+
+#### Executors of async callbacks
+
+Consider the following code:
+
+```dart
+Future<void> foo() async {
+  await something();
+  print(1);
+}
+
+void main() async {
+  await Isolate.runShared(() async {
+    await foo();
+  });
+}
+```
+
+What happens when `Future` completes in a shared isolate? Who drives event loop
+of that isolate? Which thread will callbacks run on?
+
+I propose to introduce another concept similar to `Zone`: `Executor`. Executors
+encapsulate the notion of the event loop and control how tasks are executed.
+
+```dart
+abstract interface class Executor implements Shareable {
+    /// Returns the current executor.
+    static Executor get current;
+
+    Isolate get owner;
+
+    /// Schedules the given task to run in the given executor.
+    ///
+    /// If the current isolate is not the [owner] of this executor
+    /// the behavior depends on [copy]:   ///
+    ///   * When [copy] is `false` (which is default) [task] is
+    ///     expected to be [Shareable] and is transfered to the
+    ///     target isolate directly. If it is not [Shareable] then
+    ///     `ArgumentError` will be thrown.
+    ///   * If [copy] is `true` then [task] is copied to the
+    ///     target isolate using the same algorithm [SendPort.send]
+    ///     uses.
+    ///
+    void schedule(void Function() task,
+                  {bool copy = false});
+}
+```
+
+How a particular executor runs scheduled tasks depends on the executor itself:
+e.g. an executor can have a pool of threads or notify an embedder that it has
+tasks to run and let embedder run these tasks.
+
+All built-in asynchronous primitives will make the following API guarantee: **a
+callback passed to `Future` or `Stream` APIs will be invoked using executor
+which was running the code which registered the callback.**
+
+In other words `fut1 = fut.then(cb)` is equivalent to:
+
+```dart
+final result = ShareableBox(Completer<R>());
+final callback = ShareableBox(cb);
+final executor = Executor.current;
+fut.then((v) {
+  executor.schedule(() {
+    try {
+      final r = callback.value(v);
+      result.value.complete(r);
+    } catch (e, st) {
+      result.value.completeError(e, st);
+    }
+  });
+});
+final fut1 = result.value.future;
+```
+
+> [!NOTE]
+>
+> There is a clear parallel between `Executor` and `Zone`: asynchronous
+> callbacks attached to`Stream` and `Future` are bound to the current `Zone`.
+> Original design suggested to treat `Zone` as an executor - but this obscured
+> the core of the proposal, so current version splits this into a clear separate
+> concept of `Executor`.
```

Comment:
Makes sense. A zone is a dynamic context inside an isolate, what we capture for code leaving the synchronous execution context, so we can reinstate it before running the rest of the code. An executor is the isolate context, which we capture when the code leaves the isolate, so we can reinstate it before running the rest of the code (which means putting the execution back into that isolate.)